vertx-sql-client/src/main/java/io/vertx/sqlclient/impl/RowSetImpl.java

```diff
@@ -16,56 +16,70 @@
  */
 package io.vertx.sqlclient.impl;
 
+import io.vertx.core.json.JsonObject;
 import io.vertx.sqlclient.RowIterator;
 import io.vertx.sqlclient.RowSet;
 import io.vertx.sqlclient.Row;
 
-import java.util.NoSuchElementException;
+import java.util.ArrayList;
+import java.util.Iterator;
 import java.util.function.Function;
 import java.util.stream.Collector;
 
-class RowSetImpl extends SqlResultBase<RowSet, RowSetImpl> implements RowSet {
+class RowSetImpl<R> extends SqlResultBase<RowSet<R>, RowSetImpl<R>> implements RowSet<R> {
 
-  static Collector<Row, RowSetImpl, RowSet> COLLECTOR = Collector.of(
+  static Collector<Row, RowSetImpl<Row>, RowSet<Row>> COLLECTOR = Collector.of(
     RowSetImpl::new,
     (set, row) -> {
-      if (set.head == null) {
-        set.head = set.tail = (RowInternal) row;
-      } else {
-        set.tail.setNext((RowInternal) row);;
-        set.tail = set.tail.getNext();
-      }
+      set.list.add(row);
     },
     (set1, set2) -> null, // Shall not be invoked as this is sequential
     (set) -> set
   );
 
-  static Function<RowSet, RowSetImpl> FACTORY = rs -> (RowSetImpl) rs;
+  static <R> Collector<Row, RowSetImpl<R>, RowSet<R>> mappingCollector(Function<JsonObject, R> f) {
+    return Collector.of(
+      RowSetImpl::new,
+      (set, row) -> {
+        JsonObject o = new JsonObject();
+        int size = row.size();
+        for (int idx = 0;idx < size;idx++) {
+          String key = row.getColumnName(idx);
+          Object value = row.getValue(idx);
+          o.put(key, value);
```

Comment:
I don't think this will work for non-JSON types such as `DATETIME` unless we made a transformation here.

Reply:
yes I know, actually we will update JsonObject so non json objects are stored as is in the underlying map and the `Instant` to String transformation is done when serializing to a String or a Buffer

Reply:
I'll try to provide a patch for this in vertx-core

Reply:
we should have some integration tests for this though

Reply:
I think in practice anything will be serialised for non JSON types but it should work as long as it can be serialized and deserialized. For instance it works with Instant because Vert.x has jackson serializer / deserializer for those. It might create un-necessary copies though.

Reply:
storing stuff as non string would only help for Vert.x data object mapping that would use directly the right value from JSON

Reply:
so perhaps for now we should only merge all the commits but not the last commit about mapping, given that this commit is quite easy and we give more thought at it

Reply:
what we could do is just provide mapping using Row and not JSON, i.e `Function<Row, T>` and allow anyone to have its transformation here without for now the `JsonObject`.

Reply:
> so perhaps for now we should only merge all the commits but not the last commit about mapping, given that this commit is quite easy and we give more thought at it

yes that's pragmatic.

> what we could do is just provide mapping using `Row` and not JSON, i.e `Function<Row, T>` and allow anyone to have its transformation here without for now the `JsonObject`.

Maybe we can provide methods converting `Row` to `Data Object` if we can support all Java types as proposed in https://github.com/vert-x3/wiki/wiki/RFC%3A-codegen-any-Java-type-support.

Reply:
I think for now I will only push the changes without mapping and discuss more in depth mapping itself in the community because I can see other options for this.

Reply:
Basically I don't think we need to use `JsonObject` and we could use instead the `Iterable<Map.Entry<String, Object>>` interface I think, so reusing data object would still work and we would not be coupled that much to JSON.

We would also have `Row` implement the same interface and then we don't need to use JsonObject for this

Reply:
yes that's the way we should go for now.

Reply:
For now I will only push the changes without the mapping and let this PR open as a draft for this future feature